File: src/org/elixir_lang/ElixirModules.java
package org.elixir_lang;

import com.intellij.openapi.util.io.FileUtil;
import com.intellij.util.ResourceUtil;
import com.intellij.util.io.URLUtil;
import org.elixir_lang.jps.builder.ParametersList;
import org.jetbrains.annotations.NotNull;

import java.io.*;
import java.net.URL;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;

public class ElixirModules {
    private ElixirModules() {
    }

    @NotNull
    public static ParametersList add(@NotNull ParametersList parametersList, @NotNull List<File> fileList) {
        for (File file : fileList) {
            parametersList.add("-r", file.getPath());
        }

        return parametersList;
    }

    @NotNull
    private static List<File> copy(@NotNull String basePath,
                                   @NotNull List<String> relativePathList,
                                   @NotNull String destinationDirectoryPath) throws IOException {
        List<File> destinationFileList = new ArrayList<>(relativePathList.size());

        for (String relativePath : relativePathList) {
            destinationFileList.add(copy(basePath, relativePath, destinationDirectoryPath));
        }

        return destinationFileList;
    }

    @NotNull
    private static File copy(@NotNull String basePath,
                             @NotNull String relativePath,
                             @NotNull String destinationDirectoryPath) throws IOException {
        URL moduleUrl = ResourceUtil.getResource(ElixirModules.class, basePath, relativePath);

        if (moduleUrl == null) {
            throw new IOException(
                    "Failed to locate Elixir module (under base path `" + basePath + "` on relative path `" +
                            relativePath + "`"
            );
        }

        return copy(moduleUrl, Paths.get(destinationDirectoryPath, basePath, relativePath).toFile());
    }

    @NotNull
    private static File copy(@NotNull URL moduleURL, File destination) throws IOException {
        try (BufferedInputStream inputStream = new BufferedInputStream(URLUtil.openStream(moduleURL))) {
            //noinspection ResultOfMethodCallIgnored
            destination.getParentFile().mkdirs();

            try (BufferedOutputStream outputStream = new BufferedOutputStream(new FileOutputStream(destination))) {
                FileUtil.copy(inputStream, outputStream);
            }
        }

        return destination;
    }

    @NotNull
    public static List<File> copy(@NotNull String basePath, @NotNull List<String> relativePathList) throws IOException {
        File temporaryDirectory = FileUtil.createTempDirectory("intellij_elixir", null);
        return copy(basePath, relativePathList, temporaryDirectory.getPath());
    }

    @NotNull
    public static ParametersList parametersList(@NotNull List<File> fileList) {
        return add(new ParametersList(), fileList);
    }
}

File: src/org/elixir_lang/ElixirModules.kt
package org.elixir_lang

import com.intellij.openapi.util.io.FileUtil
import com.intellij.util.ResourceUtil
import com.intellij.util.io.URLUtil
import java.io.*
import java.net.URL
import java.nio.file.Paths
import java.util.*

object ElixirModules {
    fun add(parametersList: MutableList<String>, fileList: kotlin.collections.List<File>): MutableList<String> {
        for (file in fileList) {
            parametersList.add("-r")
            parametersList.add(file.path)
        }

        return parametersList
    }

    fun toRequireList(fileList: kotlin.collections.List<File>): kotlin.collections.List<String> =
            fileList.flatMap { file ->
                listOf("-r", file.path)
            }

    @Throws(IOException::class)
    private fun copy(basePath: String,
                     relativePathList: kotlin.collections.List<String>,
                     destinationDirectoryPath: String): kotlin.collections.List<File> {
        val destinationFileList = ArrayList<File>(relativePathList.size)

        for (relativePath in relativePathList) {
            destinationFileList.add(copy(basePath, relativePath, destinationDirectoryPath))
        }

        return destinationFileList
    }

    @Throws(IOException::class)
    private fun copy(basePath: String,
                     relativePath: String,
                     destinationDirectoryPath: String): File {
        val moduleUrl = ResourceUtil.getResource(ElixirModules::class.java, basePath, relativePath)
                ?: throw IOException(
                        "Failed to locate Elixir module (under base path `" + basePath + "` on relative path `" +
                                relativePath + "`"
                )

        return copy(moduleUrl, Paths.get(destinationDirectoryPath, basePath, relativePath).toFile())
    }

    @Throws(IOException::class)
    private fun copy(moduleURL: URL, destination: File): File {
        BufferedInputStream(URLUtil.openStream(moduleURL)).use { inputStream ->

            destination.parentFile.mkdirs()

            BufferedOutputStream(FileOutputStream(destination)).use { outputStream -> FileUtil.copy(inputStream, outputStream) }
        }

        return destination
    }

    @Throws(IOException::class)
    fun copy(basePath: String, relativePathList: kotlin.collections.List<String>): kotlin.collections.List<File> {
        val temporaryDirectory = FileUtil.createTempDirectory("intellij_elixir", null)
        return copy(basePath, relativePathList, temporaryDirectory.path)
    }

    fun parametersList(fileList: kotlin.collections.List<File>): kotlin.collections.List<String> = add(mutableListOf(), fileList)
}
